_posts/2022-09-28-import-big-csv.md

```diff
@@ -0,0 +1,306 @@
+little memory but can be a little slow as a single thread is used to
+read the file. When using `open_csv()`, the data types need to be
+consistent in your columns. The inferrence is made on the first chunk,
+and if the type changes halfway through your dataset, you will run
+into errors. You can avoid this by specifying the data types manually.
+
+```python
+# from https://stackoverflow.com/a/68563617/1113276
+import pyarrow as pa
+import pyarrow.parquet as pq
+import pyarrow.csv
+
+in_path = '~/datasets/data.csv'
+out_path = '~/datasets/data.parquet'
+
+writer = None
+with pyarrow.csv.open_csv(in_path) as reader:
+    for next_chunk in reader:
+        if next_chunk is None:
+            break
+        if writer is None:
+            writer = pq.ParquetWriter(out_path, next_chunk.schema)
+        next_table = pa.Table.from_batches([next_chunk])
+        writer.write_table(next_table)
+writer.close()
+```
+
+On my system, the conversion from the CSV to Parquet takes about 190
+seconds. Reading the parquet file can be done with:
+
+```python
+data = pq.ParquetDataset(out_path).read()
+```
+
+With this approach the dataset is in memory, and for this large
+dataset, on my system, I need to be careful with what is running on my
+system (I can't have my web browser open for instance) to be able to
+load this without running out of memory.
+
+## The Dataset API in Python
+
+To load the CSV file with the Dataset API:
+
+```python
+import pyarrow.dataset as ds
+
+in_path = "~/datasets/data.csv"
+out_path = "~/datasets/my-data/"
+
+data = ds.dataset(in_path)
+```
+
+Just like with R, reading this file takes about 0.02 seconds.
+
+To convert it to parquet, you use the `write_dataset()` function:
+
+```python
+ds.write_dataset(data, out_path,
+                 format = "parquet",
+                 max_rows_per_file = 1e7)
+```
+
+And reading the collection of parquet files can also be done with the
+`dataset()` function, just like when we used the function to read the
+CSV file. The `format` argument is optional as the function detects
+automatically the file type.
+
+
+```python
+data = ds.dataset(out_path, format = "parquet")
+```
+
+## What I didn't talk about
+
+I focused on the reading of a CSV file and its conversion to
+Parquet. I didn't talk about all the options that both the Single file
+and the Dataset APIs have to customize the format of the files that
+are being imported. For instance, both APIs can be used to specify a
+different column-separator, and cell content that should be treated as
+missing data.
+
+## Conclusion
+
+Unless you are working with a relatively small dataset stored as a
+single file, using the Dataset API provides you with flexibility to
+read and write datasets quickly and in a variety of formats. The
+approach described here would work just as well with a much larger
+dataset.
+
+To summarize what we learned in this post, I put together this brief
+decision guide to help you choose the correct API to import your
```

Comment:
I'm not sure this was effectively summarized, because we never explored the advantages of single-file API in a place where it can really shine. This article more proved how good Dataset is for large files, and that the only trade-off is a one-time conversion to a Parquet dataset.